test/common/quic/envoy_quic_h3_fuzz_helper.cc

```diff
@@ -0,0 +1,431 @@
+        toStatelessResetToken(f.stateless_reset_token());
+    new_connection_id_frame.sequence_number = clampU64(f.sequence_number());
+    frames.push_back(quic::QuicFrame(&new_connection_id_frame));
+  } break;
+  case QuicFrame::kRetireConnectionId: {
+    // no retire frame in IETF mode
+    const auto& f = frame.retire_connection_id();
+    retire_connection_id_frame = std::make_unique<quic::QuicRetireConnectionIdFrame>(
+        f.control_frame_id(), f.sequence_number());
+    frames.push_back(quic::QuicFrame(retire_connection_id_frame.get()));
+
+    return;
+  } break;
+  case QuicFrame::kMaxStreams: {
+    const auto& f = frame.max_streams();
+    frames.push_back(quic::QuicFrame(
+        quic::QuicMaxStreamsFrame(f.control_frame_id(), f.stream_count(), f.unidirectional())));
+  } break;
+  case QuicFrame::kStreamsBlocked: {
+    const auto& f = frame.streams_blocked();
+    frames.push_back(quic::QuicFrame(
+        quic::QuicStreamsBlockedFrame(f.control_frame_id(), f.stream_count(), f.unidirectional())));
+  } break;
+  case QuicFrame::kPathResponse: {
+    const auto& f = frame.path_response();
+    frames.push_back(quic::QuicFrame(
+        quic::QuicPathResponseFrame(f.control_frame_id(), toPathFrameBuffer(f.data()))));
+  } break;
+  case QuicFrame::kPathChallenge: {
+    const auto& f = frame.path_challenge();
+    frames.push_back(quic::QuicFrame(
+        quic::QuicPathChallengeFrame(f.control_frame_id(), toPathFrameBuffer(f.data()))));
+  } break;
+  case QuicFrame::kStopSending: {
+    const auto& f = frame.stop_sending();
+    quic::QuicRstStreamErrorCode error_code =
+        static_cast<quic::QuicRstStreamErrorCode>(f.error_code());
+    frames.push_back(quic::QuicFrame(
+        quic::QuicStopSendingFrame(f.control_frame_id(), f.stream_id(), error_code)));
+  } break;
+  case QuicFrame::kMessageFrame: {
+    const auto& f = frame.message_frame();
+    size_t len = std::min(f.data().size(), sizeof(buffer));
+    memcpy(buffer, f.data().data(), len);
+    message_frame = std::make_unique<quic::QuicMessageFrame>(buffer, len);
+    frames.push_back(quic::QuicFrame(message_frame.get()));
+  } break;
+  case QuicFrame::kNewToken: {
+    const auto& f = frame.new_token();
+    size_t len = std::min(f.token().size(), sizeof(buffer));
+    memcpy(buffer, f.token().data(), len);
+    absl::string_view token(buffer, len);
+    new_token = std::make_unique<quic::QuicNewTokenFrame>(f.control_frame_id(), token);
+    frames.push_back(quic::QuicFrame(new_token.get()));
+  } break;
+  case QuicFrame::kAckFrequency: {
+    const auto& f = frame.ack_frequency();
+    auto delta = quic::QuicTime::Delta::FromMilliseconds(clampU64(f.milliseconds()));
+    ack_frequency = std::make_unique<quic::QuicAckFrequencyFrame>(
+        f.control_frame_id(), clampU64(f.sequence_number()), clampU64(f.packet_tolerance()), delta);
+    frames.push_back(quic::QuicFrame(ack_frequency.get()));
+  } break;
```

Comment:
This is a really long switch block. Could you move the logic into helper functions for some complicated cases?

Reply:
This is a very annoying switch case. The problem that I faced here is that Quic frames, e.g. `QuicCryptoFrame`, but also others, do not take ownership of their data, hence there is a char buffer on the stack, which guarantees the lifetime of the data until the function returns. However the constructor of  `QuicFrame` also takes a pointer to, e.g. `QuicCryptoFrame`, which prevents me to construct `QuicCryptoFrame` inside of the switch case block on the stack, because it will go out of scope before the packet is actually serialized after the switch case. Maybe a solution with minimal boilerplate is to construct a serialize closure that can be passed to handling functions...

Reply:
Didn't need to be a closure.

Reply:
Okay, thanks for the explanation about the frame lifetime. It makes sense to have those unique_ptr. But still I think you can make helper functions to return a unique_ptr of the specific frames and hold the object through out the scope of this function, i.e. unique_ptr<quic::QuicStreamFrame> serializeMessageFrame(const StreamFrameProto& proto).